students/thorn/lesson04/trigram.py

```diff
@@ -0,0 +1,118 @@
+# Thomas Horn
+# Lesson 4 - Trigram Analysis
+# March 2018
+
+################################################################################
+# Trigram Analysis: look at each set of 3 adjacent words in sherlock.txt and/or
+# sherlock_small.txt.
+#   - First two words = key -> third word = value
+#
+# This starts at an arbitrary point and looks up a random next word using the
+# table created with the above method.  This gives a new word pair at the end of
+# the text which is used to look up a new potential word.
+################################################################################
+
+import os
+import re
+import random
+
+def read_text(intext):
+    """
+    Creates the word pairs and associated new sentence words to build the
+    trigram.
+
+    Book will need to have punctuation and whitespace removed.
+    """
+    with open(intext, 'r') as infile:
+        raw_text = infile.read()
+        # Strip to only letters and spaces between words, then add to list.
+        formatted_text = re.sub(r'\W', ' ', raw_text.strip().lower())
+        text_list = list(formatted_text.split())
+    return text_list            
+
+def create_digrams(text_list):
+    """
+    FOR TESTING ONLY
+
+    Creates a list of digram tuples from the text.  
+    
+    Digram consist of the first word and the word following.
+    Ex: ([0],[0+1])
+    """
+    digrams = []
+    for i in range(len(text_list)):
+        if i < len(text_list)-1:
+            digrams.append((text_list[i], text_list[i+1]))
+    return digrams    
+
+
+def create_trigrams(text_list):
+    """
+    Creates a list of trigram tuples from the text.
+
+    Trigrams consist of the first 2 words and the word following.
+    Ex: ([0],[0+1],[0+2])
+    """
+    trigrams = []
+    for i in range(len(text_list)-2):
+        trigrams.append((text_list[i], text_list[i+1], text_list[i+2]))
+
+    # Create set and shuffle for randomness.
+    random.shuffle(trigrams)
+
+    trigram_set = set(trigrams)
```

Comment:
the trigrams created here are not exactly as per instructions on the course website:
this text -> "I wish I may I wish I might"
should produce a map where you have a key of two consecutive words as a key and values which are words that are possible to follow the key:
```
"I wish" => ["I", "I"]
"wish I" => ["may", "might"]
"may I"  => ["wish"]
"I may"  => ["I"]
```

I don't see that happening here

Reply:
Oooh, I see now.  I completely misunderstood the concept and thought that I needed to just use each word once based on the trigram value versus ALL possible words that follow the key.

I'll fix this up and resubmit.  I will admit that this was one where the output was conceptually difficult for me to grasp, so no promises :)

Reply:
🙂 lets us know if you need some help clarifying anything

essentially you should be creating a "map" of two words to a list of all possible words that follow it (hint: dictionary)